matchbox_socket/src/webrtc_socket/mod.rs

```diff
@@ -13,64 +13,76 @@ use futures_timer::Delay;
 use futures_util::select;
 use log::{debug, error, warn};
 use matchbox_protocol::PeerId;
-use messages::*;
+pub use messages::*;
 pub(crate) use socket::MessageLoopChannels;
 pub use socket::{
     ChannelConfig, PeerState, RtcIceServerConfig, WebRtcChannel, WebRtcSocket, WebRtcSocketBuilder,
 };
-use std::{collections::HashMap, pin::Pin, time::Duration};
+use std::{collections::HashMap, pin::Pin, sync::Arc, time::Duration};
 
 cfg_if! {
     if #[cfg(target_arch = "wasm32")] {
         mod wasm;
         type UseMessenger = wasm::WasmMessenger;
-        type UseSignaller = wasm::WasmSignaller;
+        type UseSignallerBuilder = wasm::WasmSignallerBuilder;
         /// A future which runs the message loop for the socket and completes
         /// when the socket closes or disconnects
         pub type MessageLoopFuture = Pin<Box<dyn Future<Output = Result<(), Error>>>>;
     } else {
         mod native;
         type UseMessenger = native::NativeMessenger;
-        type UseSignaller = native::NativeSignaller;
+        type UseSignallerBuilder = native::NativeSignallerBuilder;
         /// A future which runs the message loop for the socket and completes
         /// when the socket closes or disconnects
         pub type MessageLoopFuture = Pin<Box<dyn Future<Output = Result<(), Error>> + Send>>;
     }
 }
 
+/// A builder for a signalling implementation.
 #[cfg_attr(not(target_arch = "wasm32"), async_trait)]
 #[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
-trait Signaller: Sized {
-    async fn new(mut attempts: Option<u16>, room_url: &str) -> Result<Self, SignalingError>;
+pub trait SignallerBuilder: std::fmt::Debug + Sync + Send + 'static {
+    /// Create a new signaller.
+    async fn new_signaller(
+        &self,
+        attempts: Option<u16>,
+        room_url: String,
+    ) -> Result<Box<dyn Signaller>, SignalingError>;
+}
 
-    async fn send(&mut self, request: String) -> Result<(), SignalingError>;
+/// A signalling implementation.
+#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
+pub trait Signaller: Sync + Send + 'static {
+    /// Send a message to the signaller.
+    async fn send(&mut self, request: PeerRequest) -> Result<(), SignalingError>;
```

Comment:
When reading the match box code, I've found it a hard to know what exactly a PeerRequest is. The added doc here is nice, but I think its a bit unclear and could use more information. I suspect you figured this out as part of your project, so including some details about it would be helpful.

Specifically details on what this method is used to accomplish would be nice.

Does this initiate a request to peer with a peer indicated in the peer request? If so something like:
```suggestion
    /// Initiate an attempt to connect to the peer indicated in the [PeerRequest].
    async fn send(&mut self, request: PeerRequest) -> Result<(), SignalingError>;
```

Or maybe this requests a list of peers from the server?
Clarifying this here and linking PeerRequest and maybe adding a sentence there would be awesome.

Reply:
https://github.com/johanhelsing/matchbox/pull/485/commits/57c02f15f75bd2301854723cfac9c84c296a7ec5#diff-033c93ea2036b300064916fdfa7f7f6efe04d368435a7c845da5d01cde003883R54